lib/galaxy/tools/toolbox/cache.py

```diff
@@ -13,16 +14,39 @@ def __init__(self):
         self._hash_by_tool_paths = {}
         self._tools_by_path = {}
         self._tool_paths_by_id = {}
+        self._mod_time_by_path = {}
 
     def cleanup(self):
-        """Remove uninstalled tools from tool cache if they are not on disk anymore or if their content has changed."""
-        paths_to_cleanup = {path: tool.all_ids for path, tool in self._tools_by_path.items() if not os.path.exists(path) or md5_hash_file(path) != self._hash_by_tool_paths[path]}
-        for config_filename, tool_ids in paths_to_cleanup.items():
-            del self._hash_by_tool_paths[config_filename]
-            del self._tools_by_path[config_filename]
-            for tool_id in tool_ids:
-                if tool_id in self._tool_paths_by_id:
-                    del self._tool_paths_by_id[tool_id]
+        """
+        Remove uninstalled tools from tool cache if they are not on disk anymore or if their content has changed.
+
+        Returns list of tool_ids that have been removed.
+        """
+        try:
+            paths_to_cleanup = {path: tool.all_ids for path, tool in self._tools_by_path.items() if self._should_cleanup(path)}
+            removed_tool_ids = []
+            for config_filename, tool_ids in paths_to_cleanup.items():
+                removed_tool_ids.extend(tool_ids)
+                del self._hash_by_tool_paths[config_filename]
+                del self._tools_by_path[config_filename]
+                for tool_id in tool_ids:
+                    if tool_id in self._tool_paths_by_id:
+                        del self._tool_paths_by_id[tool_id]
+            return removed_tool_ids
+        except Exception:
+            # If by chance the file is being removed while calculating the hash or modtime
+            # we don't want the thread to die.
+            return []
+
+    def _should_cleanup(self, config_filename):
+        """Return True of `config_filename` does not exist or if modtime and hash have changes, else return False."""
+        if not os.path.exists(config_filename):
+            return True
+        new_mtime = time.ctime(os.path.getmtime(config_filename))
+        if self._mod_time_by_path.get(config_filename) != new_mtime:
```

Comment:
I was assuming it would be slightly faster to look at the mtime than to read in all tool xml files and calculate the hash on them. Though I think this takes a negligible amount of time in both cases. We do the same thing in the ToolConfWatcher (it was initially only mtime, I added the hashing to not trigger unnecessary reloads during the install process).